File: apps/web-nuxt/tests/parity/cases/key-transparency.ts
// Input cases for the key-transparency (KT) log parity vectors. Pure SHA-256
// hash chain, same construction as the commitment ledger; bless manufactures
// concrete entries so the frozen args replay with no builder.
import { computeKtEntryHash, KT_GENESIS, type KtEntry } from '#shared/key-transparency'

interface RawCase {
  fn: string
  args: unknown[]
}

const ISO1 = '2026-06-19T10:00:00.000Z'
const ISO2 = '2026-06-19T11:00:00.000Z'

async function makeKtEntry(seq: number, prevHash: string, userId: string, publicKey: string, createdAt: string): Promise<KtEntry> {
  const entryHash = await computeKtEntryHash({ seq, prevHash, userId, publicKey, createdAt })
  return { seq, prevHash, userId, publicKey, createdAt, entryHash }
}

export async function buildCases(): Promise<RawCase[]> {
  const cases: RawCase[] = []

  cases.push({
    fn: 'computeKtEntryHash',
    args: [{ seq: 0, prevHash: KT_GENESIS, userId: 'u1', publicKey: 'pk-A', createdAt: ISO1 }],
  })

  const e0 = await makeKtEntry(0, KT_GENESIS, 'u1', 'pk-A', ISO1)
  const e1 = await makeKtEntry(1, e0.entryHash, 'u2', 'pk-B', ISO2)
  const e1rot = await makeKtEntry(1, e0.entryHash, 'u1', 'pk-A2', ISO2) // u1 rotates key

  cases.push({ fn: 'verifyKtChain', args: [[]] }) // empty -> genesis head
  cases.push({ fn: 'verifyKtChain', args: [[e0, e1]] }) // valid
  cases.push({ fn: 'verifyKtChain', args: [[e1]] }) // wrong start index -> sequence
  cases.push({ fn: 'verifyKtChain', args: [[e0, { ...e1, prevHash: 'wrong' }]] }) // link
  cases.push({ fn: 'verifyKtChain', args: [[{ ...e0, entryHash: 'f'.repeat(64) }]] }) // entry-hash

  cases.push({ fn: 'loggedKeyFor', args: [[e0, e1], 'u1'] }) // present
  cases.push({ fn: 'loggedKeyFor', args: [[e0, e1rot], 'u1'] }) // rotated -> latest key
  cases.push({ fn: 'loggedKeyFor', args: [[e0, e1], 'nobody'] }) // absent -> null

  return cases
}

File: apps/mobile-flutter/parity/lib/key_transparency.dart
import 'commitment.dart' show sha256Hex;

export 'commitment.dart' show sha256Hex;

// Dart port of shared/key-transparency.ts (the chat public-key hash chain), the
// single copy the Flutter app also imports. Each entry commits to the previous
// head, so a server that ever substituted a user's key would break the chain
// (or, more subtly, show a different key than the one you verified out-of-band).
const ktGenesis = '0000000000000000000000000000000000000000000000000000000000000000';
const _ktPrefix = 'ngc-kt-v1:';

String computeKtEntryHash(Map l) =>
    sha256Hex(_ktPrefix + [l['seq'], l['prevHash'], l['userId'], l['publicKey'], l['createdAt']].join(':'));

/// Result of walking the chain: `ok` plus how far it got and where it broke.
class KtVerification {
  const KtVerification({required this.ok, required this.count, required this.head, this.failure});
  final bool ok;
  final int count;
  final String head;
  final String? failure;

  /// The shape the frozen vectors carry (`failure` absent when ok).
  Map<String, dynamic> toJson() => {
        'ok': ok,
        'count': count,
        'head': head,
        if (failure != null) 'failure': failure,
      };
}

KtVerification verifyKtChain(List entries, [String genesis = ktGenesis]) {
  var prev = genesis;
  for (var i = 0; i < entries.length; i++) {
    final e = entries[i] as Map;
    if (e['seq'] != i) return KtVerification(ok: false, count: i, head: prev, failure: 'sequence');
    if (e['prevHash'] != prev) return KtVerification(ok: false, count: i, head: prev, failure: 'link');
    final h = computeKtEntryHash(e);
    if (h != e['entryHash']) {
      return KtVerification(ok: false, count: i, head: prev, failure: 'entry-hash');
    }
    prev = h;
  }
  return KtVerification(ok: true, count: entries.length, head: prev);
}

/// The most recent logged public key for a user (null if never logged).
String? loggedKeyFor(List entries, String userId) {
  for (var i = entries.length - 1; i >= 0; i--) {
    final e = entries[i] as Map;
    if (e['userId'] == userId) return e['publicKey'] as String;
  }
  return null;
}
